Add pycapsule utility to jax.extend.ffi

andportnoy opened the pull request:
This allows one to build PyCapsules ready for XLA custom call registration out of function pointers retrieved from shared libraries using ctypes. In particular, this obviates the need to create Python bindings to custom call targets.

superbobry commented on jax/_src/ffi.py:
```diff
@@ -15,9 +15,48 @@
 
 from jax._src.lib import jaxlib
 
+# Use Python stable C API function to construct a PyCapsule.
```
Nit: this comment should probably be in the function body rather than before it.

Part of it seems to belong in the docstring, and part, I think can be removed.

andportnoy replied:
Agreed, I made the comment more concise and moved it into the function body. Didn't see anything that would fit well in the docstring.

superbobry commented on jax/_src/ffi.py:
```diff
@@ -15,9 +15,48 @@
+# Following this advice we annotate argument and return types of PyCapsule_New
+# before we call it. Example here:
+# https://stackoverflow.com/questions/65056619/converting-ctypes-c-void-p-to-pycapsule
+def build_capsule(funcptr):
```
Maybe just `pycapsule` (`py_capsule`) or `capsule`?

andportnoy replied:
Renamed to `pycapsule`.

superbobry commented on jax/_src/ffi.py:
```diff
@@ -15,9 +15,48 @@
+
+    fooso = ctypes.cdll.LoadLibrary('./foo.so')
+    xla_client.register_custom_call_target(
+      name="bar",
```
Nit: Google Python style uses 4 space hanging indent.

andportnoy replied:
Fixed, thank you.

superbobry commented:
Looks good! Please squash the commits so we can import.

andportnoy commented:
@superbobry done, thank you for reviewing.

andportnoy commented:
Let's hold off merging until after @dfm's https://github.com/google/jax/pull/21531 goes in. I'll rebase after that.

dfm commented:
Ha I was just about to propose the opposite :D

But that sounds good! Should get it in shortly.

andportnoy committed:
Add pycapsule utility to jax.extend.ffi

This allows one to build PyCapsules ready for XLA custom call
registration out of function pointers retrieved from shared libraries
using ctypes. In particular, this obviates the need to create Python
bindings to custom call targets.

andportnoy commented:
@superbobry I've rebased to update for the new location of `ffi.py`, please re-review!
